src/crypto/tls/delegated_credentials_test.go

```diff
@@ -0,0 +1,413 @@
+	//	t.Error("Delegated Credential with wrong scheme is valid; want invalid")
+	//}
+
+	delegatedCred.Algorithm = ECDSAWithP256AndSHA256
+
+	// Test delegation certificate binding
+	cert.Leaf.Raw[0] ^= byte(42)
+	if delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+		t.Error("Delegated Credential with wrong certificate is valid; want invalid")
+	}
+
+	// Test validation of DC using a certificate that can't delegate.
+	if delegatedCred.Validate(dcTestCerts["no dc"].Leaf, "server", dcNow) {
+		t.Error("Delegated Credential with non-delegation cert is valid; want invalid")
+	}
+}
+
+// Test encoding/decoding of delegated credentials.
+func TestDelegatedCredentialMarshal(t *testing.T) {
+	initialize()
+	cert := dcTestCerts["dc"]
+	time := dcNow.Sub(cert.Leaf.NotBefore) + dcMaxTTL
+
+	delegatedCred, _, err := NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, time, "server")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ser, err := delegatedCred.Marshal()
+	if err != nil {
+		t.Error(err)
+	}
+
+	delegatedCred2, err := UnmarshalDelegatedCredential(ser)
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = equal(delegatedCred, delegatedCred2)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if delegatedCred.Algorithm != delegatedCred2.Algorithm {
+		t.Errorf("scheme mismatch: got %04x; want %04x", delegatedCred2.Algorithm, delegatedCred.Algorithm)
+	}
+
+	if !bytes.Equal(delegatedCred2.Signature, delegatedCred.Signature) {
+		t.Error("Signature mismatch")
+	}
+}
+
+var dcTests = []struct {
+	clientDCSupport  bool
+	serverDCSupport  bool
+	clientSkipVerify bool
+	clientMaxVers    uint16
+	serverMaxVers    uint16
+	expectSuccess    bool
+	expectDC         bool
+	name             string
+}{
+	{true, true, false, VersionTLS13, VersionTLS13, true, true, "tls13"},
+	{true, true, true, VersionTLS13, VersionTLS13, true, true, "tls13, using server skip verify"},
+	{false, true, false, VersionTLS13, VersionTLS13, true, false, "client no DC"},
+	{true, false, false, VersionTLS13, VersionTLS13, true, false, "server no DC"},
+	{true, false, false, VersionTLS12, VersionTLS13, true, false, "client using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS12, true, false, "server using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS11, VersionTLS13, true, false, "client using TLS 1.1. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS10, false, false, "server using TLS 1.0. No DC is supported in that version."},
+}
+
+// Checks that the client suppports a version >= 1.3 and accepts delegated
+// credentials. If so, it returns the delegation certificate; otherwise it
+// returns a non-delegated certificate.
+func testServerGetCertificate(ch *ClientHelloInfo) (*Certificate, error) {
+	versOk := false
+	for _, vers := range ch.SupportedVersions {
+		versOk = versOk || (vers >= uint16(VersionTLS13))
+	}
+
+	ch.SupportDelegatedCredential = true
```

Comment:
here

Reply:
I'll have to think about this and change it appropriately ;) I'll come back to this.

Reply:
Check commit https://github.com/cloudflare/go/pull/28/commits/7b0e3dca5153b3854d093b0d10a8fcdefb489b12, which correctly inputs how it gets populated @Lekensteyn